pkg/gitops/repository.go

```diff
@@ -0,0 +1,180 @@
+	var errs []error
+	// Close all resources of temporary deploy key.
+	if keyErrs := r.sshKey.close(ctx); keyErrs != nil {
+		errs = append(errs, keyErrs...)
+	}
+	// Delete temporary repository from the local filesystem.
+	if err := os.RemoveAll(r.tmpRepoPath); err != nil {
+		errs = append(errs, fmt.Errorf("remove temporary repository: %w", err))
+	}
+	return errs
+}
+
+func (r repository) localPath() string {
+	return r.tmpRepoPath
+}
+
+func (r repository) gitClone() error {
+	_, err := r.git("clone",
+		"--branch", r.remote.Branch, "--single-branch", r.remote.URL, ".")
+	return err
+}
+
+func (r repository) workingDirectoryClean() (bool, error) {
+	status, err := r.git("status")
+	if err != nil {
+		return false, err
+	}
+	return strings.Contains(status, "nothing to commit, working tree clean"), nil
+}
+
+func (r repository) gitCheckoutNewBranch() error {
+	// Generate branch name based on the current time.
+	t := time.Now()
+	branch := fmt.Sprintf("ci-%d-%02d-%02dT%02d-%02d-%02d",
+		t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second())
+	// Execute git checkout to a new branch with that name.
+	if _, err := r.git("checkout", "-b", branch); err != nil {
+		return fmt.Errorf("checkout new branch %q: %w", branch, err)
+	}
+	return nil
+}
+
+func (r repository) gitCommitAndPush(message string) error {
+	// Stage all changes, commit them to the current branch
+	// and push the commit to the remote repository.
+	gitArgs := [][]string{
+		{"add", "--all"},
+		{"commit", "-m", message},
+		{"push", "--all", "-u"},
+	}
+	for _, a := range gitArgs {
+		if _, err := r.git(a...); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r repository) currentBranch() (string, error) {
+	branch, err := r.git("rev-parse", "--abbrev-ref", "HEAD")
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(branch), nil
+}
+
+func (r repository) git(args ...string) (string, error) {
+	// Change current directory to the repositorys local clone.
+	originalDir, err := os.Getwd()
+	if err != nil {
+		return "", fmt.Errorf("get current dir: %w", err)
+	}
+	if err := os.Chdir(r.tmpRepoPath); err != nil {
+		return "", fmt.Errorf("change dir to %q: %w", r.tmpRepoPath, err)
+	}
+	// Defer a revert of the current directory to the original one.
+	defer os.Chdir(originalDir)
+
+	cmd := exec.Command("git", args...)
+
+	// Specify SSH key for git commands via environment variable.
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, fmt.Sprintf(
+		"GIT_SSH_COMMAND=ssh -i %s -o IdentitiesOnly=yes",
+		r.sshKey.privateKeyPath(),
+	))
+
+	// Run git command and returns it's combined output of stdout and stderr.
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", fmt.Errorf("run command %v: %w (output: %s)", args, err, out)
+	}
+	return string(out), nil
+}
+
+func (r repository) openPullRequest(ctx context.Context, title, body string) (string, error) {
+	// PR will be open from the current branch.
+	currBranch, err := r.currentBranch()
```

Comment:
Shouldn't we open the PR from the deploy branch?

Reply:
ArgoCD is monitoring the deploy branch for changes, therefore that's the destination (`base`) of the pull request. The current (created for config change) branch is going to be the `head` of the PR.